### PR TITLE
Fix generated code order when using TypeScript, ES6 and decorators

### PR DESCRIPTION
**Pull request by thewilkybarkid**
# ↪️ Pull Request

This fixes the problem I mentioned in https://github.com/parcel-bundler/parcel/issues/6676#issuecomment-949504559. I'm not sure I understand the actual problem, but it's something to do with the hoisting difference between `var` and `let`, and decorators. (I didn't manage to recreate it without decorators.)

Moving the call to `parcelHelpers.export()` to the end (i.e. after the `let`) seems to resolve the problem.

## 💻 Examples

I've managed to reproduce the problem in https://github.com/thewilkybarkid/parcel-access-before-initialization, where you can see the compilation result and the failure. (It's basically the same as the test case in the PR.)

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
Only ran a subset of tests locally (as they take so long); will investigate those.

**Comment by basher** [comment]
Thanks @thewilkybarkid for doing this.

Hopefully it will fix [#6676](https://github.com/parcel-bundler/parcel/issues/6676) so I can finally upgrade to Parcel 2 and still support certain legacy browsers!

**Comment by basher** [comment]
Thanks @thewilkybarkid for doing this.

Hopefully is will finally fix [#6676](https://github.com/parcel-bundler/parcel/issues/6676).

**Comment by mischnic** [comment]
Not sure if this entirely safe to do, circular imports might be problematic and also:

https://github.com/parcel-bundler/parcel/pull/7016:

> This updates our helpers to skip keys that are already defined when performing a wildcard re-export. This is safe because we hoist named exports to the top of the module, so by the time the helper is called, all named exports are already defined.

**Comment by thewilkybarkid** [comment]
Are you able to write a test that breaks?

**Comment by thewilkybarkid** [comment]
@mischnic As far as I'm aware the export order is still fine (it's just at the end of the module). The test I added contains circular imports too, but maybe you had a different idea.

I've just tried this out on a real project and it works fine, but still has the same problem with scope hoisting enabled (as the order is changed back)...

**Comment by mischnic** [comment]
That just came to mind because we had problems with this in the past as well. I haven't had time yet to try and test such a situation

File: packages/core/integration-tests/test/typescript-tsc.js [hunk 1]
```diff
@@ -56,4 +56,19 @@ describe('typescript tsc', function () {
...
+        config: path.join(
+          __dirname,
+          '/integration/typescript-es6-decorators/.parcelrc',
+        ),
+      },
+    );
+
+    const output = await run(b);
+    assert.strictEqual(output, 'ClassOne');
+  });
```
Comment (devongovett) [inline packages/core/integration-tests/test/typescript-tsc.js:73]: Should we also run this test case with scope hoisting enabled? We should verify that the code behaves the same way in dev and prod.

**Comment by devongovett** [comment]
Here's an example:

a.mjs
```js
import { c } from "./b.mjs";
console.log(c);
```

b.mjs
```js
export const foo = "foo";
export { c } from "./c.mjs";
```

c.mjs
```js
import { foo } from "./b.mjs";
export const c = "c:" + foo;
```

When run, this should throw "Cannot access 'foo' before initialization" (as it does when running the modules natively in Node or browsers). However, after applying this patch, it logs "c:undefined".

**Comment by thewilkybarkid** [comment]
Thanks @devongovett, I've opened #7350 with that test case. I'll move this back to draft and will keep looking.

**Comment by thewilkybarkid** [comment]
I've just been able to find https://github.com/angular/angular/commit/401ef71ae5b01be95d124184a0b6936fc453a5d4 and https://github.com/microsoft/TypeScript/issues/27519 which seem to describe this problem: it's a TypeScript limitation. I'm not sure how it was working in Parcel `2.0.0-beta1`, but I'll look more thoroughly at the generated output there.

Not sure what to do with this problem now though.

(The project where I've had this problem is inherited, and I've never used decorators before as they are known to be experimental. This gives me the excuse to remove them.)

**Comment by devongovett** [comment]
I think the reason it may have "worked" in a previous beta is because we weren't spec compliant yet. Are you sure this circular dependency works with normal TSC or Babel?

**Comment by thewilkybarkid** [comment]
> Are you sure this circular dependency works with normal TSC or Babel?

Seems to work with TSC: I've expanded https://github.com/thewilkybarkid/parcel-access-before-initialization to show a failure with Parcel but success with TSC.

**Pull request closed without merging**